feat(artifacts): Support embedded/base64 artifact type

sbwsg opened the pull request:
fixes https://github.com/spinnaker/spinnaker/issues/2526

![default_artifact_base64](https://user-images.githubusercontent.com/34253460/37614336-e710c746-2b80-11e8-95a0-1a12d19b0433.png)

With invalid input:

![base64_latin1_error](https://user-images.githubusercontent.com/34253460/37614340-ecb23428-2b80-11e8-835d-f7b328f3f7d6.png)

danielpeach commented on app/scripts/modules/core/src/pipeline/config/triggers/artifacts/artifact.component.ts:
```diff
@@ -52,7 +52,7 @@ class ArtifactCtrl implements IController {
     if (artifactKindConfig.length) {
       const config = artifactKindConfig[0];
-      const template: Element = config.template as any;
+      const template: string = config.template as any;
```
Isn't the type of `template` on `IArtifactKindConfig` a `string`? Should be able to drop `as any`.

danielpeach commented on app/scripts/modules/core/src/pipeline/config/triggers/artifacts/base64/base64.artifact.ts:
```diff
@@ -0,0 +1,87 @@
+             ng-model="ctrl.artifact.name" />
+    </div>
+  </div>
+  <div ng-if="ctrl.encodeDecodeError != null" class="form-group row">
```
In most of the spots where we surface this kind of error (see the Pipeline JSON edit modal), we put the error below the text box.

danielpeach commented:
Nice. 

- Is there any reasonable way to drop the "Failed to execute 'btoa' on 'Window'" part of the error message without parsing the error message? Users probably don't care how we converted their string to base64.
- Someone can disagree with me, but I wonder how useful the Base64 text box is - it's probably just noise in most cases, and users might be better served by a tooltip and/or a copy-to-clipboard icon.

sbwsg commented on app/scripts/modules/core/src/pipeline/config/triggers/artifacts/base64/base64.artifact.ts:
```diff
@@ -0,0 +1,87 @@
+
+import './base64.artifact.less';
+
+export const BASE64_ARTIFACT = 'spinnaker.core.pipeline.trigger.github.artifact';
```
whoops, thankyou!

sbwsg commented:
@ezimanyi yes indeed, that work was already done here: https://github.com/spinnaker/clouddriver/pull/2416/files

@dpeach for 1) I've added an explicit error message for when a char is outside the latin1 range so that it wont have the "Failed to executed 'btoa' on 'window'" part.  I expect that will be the most common error when converting between base64 / decoded content.  All other errors will be printed to screen with the entire error message.  Short of parsing it I can't think of a way to remove that string prefix.  And for 2) I agree about the lack of utility of the base64 text content being in the UI.  I've replaced the `<pre>` with a copy-to-clipboard icon.

sbwsg commented:
@lwander I've enabled the artifact as a matchArtifact.

lwander commented on app/scripts/modules/core/src/pipeline/config/triggers/artifacts/base64/base64.artifact.ts:
```diff
@@ -0,0 +1,46 @@
+  </div>
+  <div ng-if="ctrl.encodeDecodeError" class="form-group row">
+    <div class="col-md-12 error-message">
+      Error encoding/decoding artifact content: {{ ctrl.encodeDecodeError }}
```
Why does the match artifact have encoding errors?
